smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,147 @@
+        }
+        writer.write("public val #L: #P = #L,", param.name, param.type, param.defaultLiteral)
+    }
+
+    private fun renderCompanionObject() {
+        writer.withBlock("public companion object {", "}") {
+            write("public operator fun invoke(block: Builder.() -> Unit): EndpointParameters = Builder().apply(block).build()")
+        }
+    }
+
+    private fun renderInit() {
+        writer.withBlock("init {", "}") {
+            params.forEach {
+                if (it.isRequired) {
+                    write("""require(#1L != null) { "endpoint provider parameter #1L is required" }""", it.name)
+                }
+            }
+        }
+    }
+
+    private fun renderBuilder() {
+        writer.withBlock("public class Builder internal constructor() {", "}") {
+            params.forEach {
+                if (it.documentation != null) {
+                    dokka {
+                        write("#L", it.documentation)
+                    }
+                }
+                if (it.deprecated != null) {
+                    it.deprecated.writeKotlinAnnotation(writer)
+                }
+                write("public var #L: #P = #L", it.name, it.type, it.defaultLiteral)
+                write("")
+            }
+            withBlock("public fun build(): EndpointParameters {", "}") {
+                withBlock("return EndpointParameters(", ")") {
+                    params.forEach {
+                        if (it.deprecated != null) {
+                            writeInline("@Suppress(#S)", "DEPRECATION")
+                        }
+                        write("#L,", it.name)
+                    }
+                }
+            }
+        }
+    }
+}
+
+// kotlin-mapped representation of an endpoint parameter
+private data class KotlinEndpointParameter(
+    val name: String,
+    val type: Symbol,
+    val isRequired: Boolean,
+    // All endpoint params are nullable. Standard symbol rendering suppresses the default value for boxed types, but
+    // endpoint params do not, so we must store the default ourselves to render.
+    val defaultLiteral: String,
+    val documentation: String?,
+    val deprecated: Parameter.Deprecated?,
+)
+
+private fun Identifier.toKotlin(): String =
+    asString().replaceFirstChar(Char::lowercase)
+
+private fun ParameterType.toSymbol(): Symbol =
+    buildSymbol {
+        namespace = "kotlin"
+        name = when (this@toSymbol) {
+            ParameterType.STRING -> "String"
+            ParameterType.BOOLEAN -> "Boolean"
+        }
+    }
+
+private fun Value.toKotlinLiteral(): String =
+    when (this) {
+        is Value.Str -> expectString().doubleQuote()
+        is Value.Bool -> if (expectBool()) "true" else "false"
+        else -> throw IllegalArgumentException("unrecognized parameter value type")
+    }
+
+private fun Parameter.Deprecated.writeKotlinAnnotation(writer: KotlinWriter) =
+    writer.write("@Deprecated(#S)", message ?: "")
```

Comment:
**Comment**: Let's add a basic deprecation string in the event of a null or blank message on the parameter itself. Something like "This field is deprecated and no longer recommended for use."